storage_scrubber/src/checks.rs

```diff
@@ -527,3 +531,136 @@ async fn list_timeline_blobs_impl(
 }
+
+pub(crate) struct RemoteTenantManifestInfo {
+    pub(crate) latest_generation: Option<Generation>,
+    pub(crate) manifests: Vec<(Generation, ListingObject)>,
+    #[allow(dead_code)]
+    pub(crate) unknown_keys: Vec<ListingObject>,
+}
+
+pub(crate) struct ListTenantManifestResult {
+    pub(crate) errors: Vec<String>,
+    pub(crate) manifest_info: RemoteTenantManifestInfo,
+}
+
+/// Returns [`ListTimelineBlobsResult::MissingIndexPart`] if blob data has layer files
+/// but is missing [`IndexPart`], otherwise returns [`ListTimelineBlobsResult::Ready`].
+pub(crate) async fn list_tenant_manifests(
+    remote_client: &GenericRemoteStorage,
+    tenant_id: TenantShardId,
+    root_target: &RootTarget,
+) -> anyhow::Result<ListTenantManifestResult> {
+    let mut errors = Vec::new();
+    let mut unknown_keys = Vec::new();
+
+    let mut tenant_root_target = root_target.tenant_root(&tenant_id);
+    const TENANT_MANIFEST_STEM: &str = "tenant-manifest";
+    let original_prefix = tenant_root_target.prefix_in_bucket.clone();
+    tenant_root_target.prefix_in_bucket += TENANT_MANIFEST_STEM;
+    tenant_root_target.delimiter = String::new();
+
+    let mut manifests: Vec<(Generation, ListingObject)> = Vec::new();
+
+    let prefix_str = &original_prefix
+        .strip_prefix("/")
+        .unwrap_or(&original_prefix);
+
+    let mut stream = std::pin::pin!(stream_listing(remote_client, &tenant_root_target));
+    'outer: while let Some(obj) = stream.next().await {
+        let (key, Some(obj)) = obj? else {
+            panic!("ListingObject not specified");
+        };
+
+        let blob_name = key.get_path().as_str().strip_prefix(prefix_str);
+        'err: {
+            // TODO a let chain would be nicer here.
+            let Some(name) = blob_name else {
+                break 'err;
+            };
+            if !name.starts_with("tenant-manifest") {
+                break 'err;
+            }
+            let Ok(path) = RemotePath::from_string(name) else {
+                break 'err;
+            };
+            let Some(generation) = parse_remote_tenant_manifest_path(path) else {
+                break 'err;
+            };
+            tracing::debug!("tenant manifest {key}");
+            manifests.push((generation, obj));
+            continue 'outer;
+        }
+        tracing::info!("Listed an unknown key: {key}");
+        unknown_keys.push(obj);
+    }
+
+    if manifests.is_empty() {
+        tracing::debug!("No manifest for timeline.");
+
+        return Ok(ListTenantManifestResult {
+            errors,
+            manifest_info: RemoteTenantManifestInfo {
+                latest_generation: None,
+                manifests,
+                unknown_keys,
+            },
+        });
+    }
+
+    // Find the manifest with the highest generation
+    let (latest_generation, latest_listing_object) = manifests
+        .iter()
+        .max_by_key(|i| i.0)
+        .map(|(g, obj)| (*g, obj.clone()))
+        .unwrap();
```

Comment:
If there was any error (`!unknown_keys.is_empty()`), then we must not assume we actually see all the manifests.
We should bail with some sort of warning or error instead and not do anything.

Reply:
done